wait/awaitility.go

```diff
@@ -0,0 +1,172 @@
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kubefed/pkg/apis/core/common"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/controller/util"
+	"testing"
+	"time"
+)
+
+const (
+	OperatorRetryInterval = time.Second * 5
+	OperatorTimeout       = time.Second * 60
+	RetryInterval         = time.Millisecond * 100
+	Timeout               = time.Second * 3
+	MemberNsVar           = "MEMBER_NS"
+	HostNsVar             = "HOST_NS"
+)
+
+// Awaitility contains information necessary for verifying availability of resources in both operators
+type Awaitility struct {
+	T                *testing.T
+	Client           framework.FrameworkClient
+	ControllerClient client.Client
+	KubeClient       kubernetes.Interface
+	MemberNs         string
+	HostNs           string
+}
+
+// SingleAwaitility contains information necessary for verifying availability of resources in a single operator
+type SingleAwaitility interface {
+	WaitForKubeFedCluster(clusterType cluster.Type, condition *v1beta1.ClusterCondition) error
+	WaitForKubeFedClusterConditionWithName(name string, condition *v1beta1.ClusterCondition) error
+	GetKubeFedCluster(clusterType cluster.Type, condition *v1beta1.ClusterCondition) (v1beta1.KubeFedCluster, bool, error)
+	NewKubeFedCluster(name string, caBundle []byte, apiEndpoint, secretRef string, labels map[string]string) *v1beta1.KubeFedCluster
+}
+
+type SingleAwaitilityImpl struct {
+	T               *testing.T
+	Client          framework.FrameworkClient
+	Ns              string
+	OtherOperatorNs string
+}
+
+// Member creates SingleAwaitility for the member operator
+func (a *Awaitility) Member() *MemberAwaitility {
+	return NewMemberAwaitility(a)
+}
+
+// Host creates SingleAwaitility for the host operator
+func (a *Awaitility) Host() *HostAwaitility {
+	return NewHostAwaitility(a)
+}
+
+// ReadyKubeFedCluster is a ClusterCondition that represents cluster that is ready
+var ReadyKubeFedCluster = &v1beta1.ClusterCondition{
+	Type:   common.ClusterReady,
+	Status: v1.ConditionTrue,
+}
+
+// WaitForReadyKubeFedClusters waits until both KubeFedClusters (host and member) exist and has ready ClusterCondition
+func (a *Awaitility) WaitForReadyKubeFedClusters() error {
+	if err := a.Host().WaitForKubeFedCluster(cluster.Member, ReadyKubeFedCluster); err != nil {
+		return err
+	}
+	if err := a.Member().WaitForKubeFedCluster(cluster.Host, ReadyKubeFedCluster); err != nil {
+		return err
+	}
+	return nil
+}
+
+// WaitForKubeFedCluster waits until there is a KubeFedCluster representing a operator of the given type
+// and running in the given expected namespace. If the given condition is not nil, then it also checks
+// if the CR has the ClusterCondition
+func (a *SingleAwaitilityImpl) WaitForKubeFedCluster(clusterType cluster.Type, condition *v1beta1.ClusterCondition) error {
+	timeout := Timeout
+	if condition != nil {
+		timeout = (util.DefaultClusterHealthCheckPeriod + 5) * time.Second
```

Comment:
maybe this can be a const so we don't need to find the other place(s) it is used - rather just one place..

Reply:
good idea - done e392402